web/src/app/forms/FormField.js

```diff
@@ -9,206 +8,106 @@ import shrinkWorkaround from '../util/shrinkWorkaround'
-    return this.props.validate(value)
-  }
-
-  render() {
-    return (
-      <FormContainerContext.Consumer>
-        {this.renderComponent}
-      </FormContainerContext.Consumer>
-    )
-  }
-
-  renderComponent = ({
-    errors,
-    value,
-    onChange,
-    addField,
-    disabled: containerDisabled,
-    optionalLabels,
-    ...otherFormProps
-  }) => {
-    const {
-      errorName,
-      name,
-      noError,
-      component: Component,
-      render,
-      fieldName: _fieldName,
-      formLabel,
-      required,
-      validate,
-      disabled: fieldDisabled,
-      hint,
-      label: _label,
-      InputLabelProps: _inputProps,
-      mapValue,
-      mapOnChangeValue,
-      min,
-      max,
-      checkbox,
-      float,
-      ...otherFieldProps
-    } = this.props
-
-    const baseLabel = typeof _label === 'string' ? _label : startCase(name)
-    const label =
-      !required && optionalLabels ? baseLabel + ' (optional)' : baseLabel
-
-    const fieldName = _fieldName || name
-    const props = {
-      ...otherFormProps,
-      ...otherFieldProps,
-      name,
-      required,
-      disabled: containerDisabled || fieldDisabled,
-      error: errors.find((err) => err.field === (errorName || fieldName)),
-      hint,
-      value: mapValue(get(value, fieldName), value),
-      min,
-      max,
-      float,
-    }
-
-    const InputLabelProps = {
-      required: required && !optionalLabels,
-      ...shrinkWorkaround(props.value),
-      ..._inputProps,
-    }
-
-    let getValueOf = (e) => (e && e.target ? e.target.value : e)
-    if (checkbox) {
-      props.checked = props.value
-      props.value = props.value ? 'true' : 'false'
-      getValueOf = (e) => e.target.checked
-    } else if (otherFieldProps.type === 'number') {
-      props.label = label
-      props.value = props.value.toString()
-      props.InputLabelProps = InputLabelProps
-      getValueOf = (e) =>
-        float ? parseFloat(e.target.value) : parseInt(e.target.value, 10)
-    } else {
-      props.label = label
-      props.InputLabelProps = InputLabelProps
-    }
-
-    props.onChange = (_value) => {
-      let newValue = getValueOf(_value)
-      if (props.type === 'number' && typeof props.min === 'number')
-        newValue = Math.max(props.min, newValue)
-      if (props.type === 'number' && typeof props.max === 'number')
-        newValue = Math.min(props.max, newValue)
-      onChange(fieldName, mapOnChangeValue(newValue, value))
-    }
-
-    return (
-      <MountWatcher
-        onMount={() => {
-          this._unregister = addField(fieldName, this.validate)
-        }}
-        onUnmount={() => {
-          this._unregister()
-        }}
-      >
-        {this.renderContent(props)}
-      </MountWatcher>
-    )
+    return validate(value)
   }
 
-  renderContent(props) {
-    const { checkbox, component, formLabel, label, noError, render } =
-      this.props
+  useEffect(() => {
+    return addField(fieldName, validateField)
+  }, [])
```

Comment:
Can we raise this (and `validateField` it uses) up higher? Putting hooks as close to the top is helpful to avoid accidentally making them conditional (they should never be behind an `if` or early `return`) as code changes over time.